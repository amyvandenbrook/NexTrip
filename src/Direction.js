import React from 'react';

export const ChooseDirection = (props) => {
    if (props.directionLoaded && (window.location.href.toString().includes('#Directions') || props.testDirection)) {
        return (
            <React.Fragment>
                <div className="description">Choose a direction:
                </div>
                <input
                    type='radio'
                    name='Direction'
                    id={props.directionData0.Value}
                    value={props.directionData0.Value}
                    onClick={props.directionHandler}
                    defaultChecked></input>
                <label htmlFor={props.directionData0.Value}>{props
                        .directionData0
                        .Text
                        .toLowerCase()}</label>
                <input
                    type='radio'
                    name='Direction'
                    id={props.directionData1.Value}
                    value={props.directionData1.Value}
                    onClick={props.directionHandler}></input>
                <label htmlFor={props.directionData1.Value}>{props
                        .directionData1
                        .Text
                        .toLowerCase()}</label>
            </React.Fragment>
        )
    } else {
        return null
    }
}