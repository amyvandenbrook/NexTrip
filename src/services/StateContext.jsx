import React, { createContext, useReducer } from 'react'
import Reducer from './Reducer.jsx'

const StateContext = createContext()

const StateProvider = ({ children, initialState }) => {
    const defaultState = {
        routeData: [],
        route: undefined,
        routeChanged: false,
        directionData: [],
        direction: '',
        stopData: [],
    }
    const [state, dispatch] = useReducer(Reducer, {
        ...defaultState,
        ...initialState,
    })

    return (
        <StateContext.Provider
            value={{
                ...state,
                dispatch,
            }}
        >
            {children}
        </StateContext.Provider>
    )
}

export { StateContext, StateProvider }
