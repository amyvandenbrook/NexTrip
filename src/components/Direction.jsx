import React, { useCallback, useContext, useEffect } from 'react'
import { StateContext } from '../services/StateContext.jsx'
import { getDirections } from '../services/AxiosService.jsx'

const Direction = (props) => {
    const { route, directionData, dispatch } = useContext(StateContext)

    useEffect(() => {
        if (route) {
            getDirections(route).then((result) => {
                dispatch({ type: 'set_direction_data', payload: result.data })
            })
        }
    }, [dispatch, route])

    const setDirection = useCallback(
        (direction) => {
            dispatch({ type: 'set_direction', payload: direction })
        },
        [dispatch]
    )

    useEffect(() => {
        if (directionData.length > 0) {
            setDirection(directionData[0].Value)
        }
    }, [dispatch, directionData, setDirection])

    return directionData?.length > 0 ? (
        <div className="directions_container">
            <p className="directions_heading"> Choose a direction : </p>
            {directionData.map((direction, index) => {
                return (
                    <React.Fragment key={index}>
                        <input
                            className="directions_input"
                            type="radio"
                            name="Direction"
                            id={direction.Value}
                            value={direction.Value}
                            onClick={() => {
                                setDirection(direction.Value)
                            }}
                            defaultChecked={index === 0}
                        ></input>
                        <label
                            htmlFor={direction.Value}
                            className="directions_label"
                        >
                            {direction.Text.toLowerCase()}
                        </label>
                    </React.Fragment>
                )
            })}
        </div>
    ) : null
}

export default Direction
