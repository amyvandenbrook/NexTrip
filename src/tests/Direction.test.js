import React from 'react'
import { act, create } from 'react-test-renderer'
import { render, fireEvent, screen } from '@testing-library/react'
import axios from 'axios'
import { StateProvider } from '../services/StateContext.jsx'
import Direction from '../components/Direction.jsx'

jest.mock('axios')
axios.get = jest.fn()

describe('Direction', () => {
    let direction
    let component

    test('requests and displays directions if route exists', async () => {
        const mockReturnedData = [{ Text: 'mockDirection', Value: 'mockValue' }]
        axios.get.mockImplementationOnce(() =>
            Promise.resolve({
                status: 200,
                data: mockReturnedData,
            })
        )
        await act(() => {
            component = create(
                <StateProvider initialState={{ route: 'mockRoute' }}>
                    <Direction />
                </StateProvider>
            )
        })
        direction = component.toJSON()
        expect(direction).toMatchSnapshot()
    })

    test('direction options appear when route exists', async () => {
        const mockReturnedData = [
            { Text: 'mockDirection1', Value: 'mockValue1' },
            { Text: 'mockDirection2', Value: 'mockValue2' },
        ]
        axios.get.mockImplementationOnce(() =>
            Promise.resolve({
                status: 200,
                data: mockReturnedData,
            })
        )

        await render(
            <StateProvider initialState={{ route: 'mockRoute' }}>
                <Direction />
            </StateProvider>
        )
        const directionOption = screen.getByLabelText(
            mockReturnedData[1].Text.toLowerCase()
        )

        fireEvent.click(directionOption)

        expect(
            screen.getByLabelText(mockReturnedData[1].Text.toLowerCase()).id
        ).toStrictEqual(mockReturnedData[1].Value)
    })
})
