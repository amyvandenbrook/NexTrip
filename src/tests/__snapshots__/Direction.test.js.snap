// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Direction requests and displays directions if route exists 1`] = `
<div
  className="directions_container"
  id="Directions"
>
  <p
    className="directions_heading"
  >
     Choose a direction : 
  </p>
  <input
    className="directions_input"
    defaultChecked={true}
    id="mockValue"
    name="Direction"
    onClick={[Function]}
    type="radio"
    value="mockValue"
  />
  <label
    className="directions_label"
    htmlFor="mockValue"
  >
    mockdirection
  </label>
</div>
`;
